mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala

```diff
@@ -317,7 +318,12 @@ final class OneVsRest @Since("1.4.0") (
     val numClasses = MetadataUtils.getNumClasses(labelSchema).fold(computeNumClasses())(identity)
     instr.logNumClasses(numClasses)
 
-    val multiclassLabeled = dataset.select($(labelCol), $(featuresCol))
+    val multiclassLabeled = getClassifier match {
+      // SPARK-21306: cache weightCol if necessary
+      case c: HasWeightCol if c.isDefined(c.weightCol) && c.getWeightCol.nonEmpty =>
+        dataset.select($(labelCol), $(featuresCol), c.getWeightCol)
+      case _ => dataset.select($(labelCol), $(featuresCol))
+    }
```

Comment:
@facaiy It doesn't matter. If the classifier doesn't inherit from ```HasWeightCol```, we don't run ```setWeightCol``` for that classifier instead of printing out warning log to say ```weightCol``` doesn't take effect. You can refer [these lines of code](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala#L338) to learn how ```featuresCol``` be handled. Thanks.